# -*- coding: utf-8 -*-
"""Final_project.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/1izWF3wHDjUSLvgha1m6HaePPKs-l8HOB

https://www.pnas.org/doi/full/10.1073/pnas.1419161111

# Data Download - Rakshya
# Align the reads, Identify AS events, analyze CIGAR strings - Rakshya
# Statistical Analysis (Standard Deviation, Coorelation ....) - Gillian, Rakshya and Vrishab
# Visualization using Matplotlib - Vrishab
# Output of the data and documentation - Kristel

# BME160 Final Project
"""

!pip install biopython
!pip install pysam
!pip install pandas
!pip install numpy
!pip install scipy
!pip install pybedtools

"""# How to download RNA seq data from [NCBI SRA](https://www.ncbi.nlm.nih.gov/sra)

More detailed explanation here : https://www.ncbi.nlm.nih.gov/sra/docs/sradownload/

Brief Description

1.   Click on the link https://www.ncbi.nlm.nih.gov/sra
2.   Then on the search bar type RNA-seq of .....(I typed RNA-Seq of Colorectal Cancer)
3.   Click on the file you want to download
4.   After that copy the Run ID 
5.   data below 5gb can be downloaded from here https://www.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser&display=metadata
6.   data greater than 5gb needs to be downloaded from SRA toolkit

SRA toolkit installation steps in MacOS terminal

```zsh
$curl --output sratoolkit.tar.gz https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/current/sratoolkit.current-mac64.tar.gz

$tar -vxzf sratoolkit.tar.gz

$export PATH=$PATH:/Users/rausharm/BME160/sratoolkit.3.0.5-mac64/bin\n

$source ~/.zshrc

$/Users/rausharm/BME160/sratoolkit.3.0.5-mac64/bin/vdb-config -i

$/Users/rausharm/BME160/sratoolkit.3.0.5-mac64/bin/fastq-dump

$which fastq-dump

$fastq-dump --stdout -X 2 SRR390728
```
OUTOUT
```zsh
Read 2 spots for SRR390728
Written 2 spots for SRR390728
@SRR390728.1 1 length=72
CATTCTTCACGTAGTTCTCGAGCCTTGGTTTTCAGCGATGGAGAATGACTTTGACAAGCTGAGAGAAGNTNC
+SRR390728.1 1 length=72
;;;;;;;;;;;;;;;;;;;;;;;;;;;9;;665142;;;;;;;;;;;;;;;;;;;;;;;;;;;;;96&&&&(
@SRR390728.2 2 length=72
AAGTAGGTCTCGTCTGTGTTTTCTACGAGCTTGTGTTCCAGCTGACCCACTCCCTGGGTGGGGGGACTGGGT
+SRR390728.2 2 length=72
;;;;;;;;;;;;;;;;;4;;;;3;393.1+4&&5&&;;;;;;;;;;;;;;;;;;;;;<9;<;;;;;464262
```

Data downloaded: Prostate Cancer
https://www.ncbi.nlm.nih.gov/sra/SRX20484591[accn]

Run Id: SRR24705928

Run Browser for data:https://www.ncbi.nlm.nih.gov/Traces/index.html?view=run_browser

# Align it with the reference genome
https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.40

Downloaded both Genomic Fasta and RNA fasta 

- Easiest way to aligh might be using online resources like https://usegalaxy.org
- Encountered an error because the genome indexing process requires more RAM than has been allocated. So, using the built-in GRhg38.
- Sucessfully aligned it to the hg38 and for the files. 
- Steps followed:
    - Click the link https://usegalaxy.org and in the search bar type " RNA STAR"
    - Then upload the RNA-seq prostate cancer data 
    - Choose the built-in reference file
    - Chose the default settings for now
    - Press Run tool
    - Wait for the taks to complete
    - The run created junctions.bed and bam file.

The code below is reading the output generated after aligning them to reference sequence using pysam and pandas dataframe. I have printed first ten lines as its a huge data and also printed the junction bed files.
"""

import pysam
import pandas as pd

bamfile = pysam.AlignmentFile("/content/test.bam", "rb")

for i, read in enumerate(bamfile):
    if i >= 10:
      break
    print("Read name:", read.query_name)
    print("Reference start:", read.reference_start)
    print("Reference end:", read.reference_end)


bedfile = pd.read_table("/content/test.bed", header=None)

# Print the DataFrame
print("printing bedfile",bedfile)

"""# Identify Alternative Splicing Events
The code below analyze a BAM file containing RNA-seq data and determine the number of reads that support alternative splicing. Alternative splicing is a process in which different combinations of exons are included or excluded from the final mRNA transcript, leading to the generation of multiple protein isoforms from a single gene.

The code first defines a function detect_alternative_splicing that checks if a read spans multiple exons, indicating the presence of alternative splicing. It then opens the BAM file and iterates over each read. For each read, the code calls the detect_alternative_splicing function and increments counters for the total number of reads and the number of reads supporting alternative splicing.

The output of the code is the total number of reads and the number of reads that exhibit alternative splicing events. We get occurrence of alternative splicing events


"""

def detect_alternative_splicing(read):
    blocks = read.get_blocks()

    # Check if the read starts and ends in different blocks
    if len(blocks) > 1:
        return True
    else:
        return False

bamfile = pysam.AlignmentFile("/content/test.bam", "rb")

total_reads = 0
alternative_splicing_events = 0

for read in bamfile:
    # Check if the read supports alternative splicing
    if detect_alternative_splicing(read):
        alternative_splicing_events += 1
    total_reads += 1

print("Total Reads:", total_reads)
print("Alternative Splicing Events:", alternative_splicing_events)

bamfile.close()

"""There are a total of 12,559,327 reads in the dataset. Among these reads, there are 3,968,412 alternative splicing events detected.

# Visualizing Alternative Splicing
"""

import matplotlib.pyplot as plt

total_reads = 12559327
alternative_splicing_events = 3968412

percentage = (alternative_splicing_events / total_reads) * 100

x = ['Total Reads', 'Alternative Splicing Events']
y = [total_reads, alternative_splicing_events]

plt.bar(x, y)
plt.ylabel('Count')
plt.title('Alternative Splicing Events')
plt.text(0, total_reads, f'Total Reads: {total_reads}', ha='center')
plt.text(1, alternative_splicing_events, f'Alternative Splicing Events: {alternative_splicing_events}\n({percentage:.2f}%)', ha='center')

plt.show()

"""# Analyzing CIGAR Strings in RNA-seq Data

The code reads a BAM file containing the alignment information of the RNA-seq reads. It iterates over each read in the file and checks if it is aligned and contains splicing events. It extracts the CIGAR string from the read, which represents the alignment and possible gaps/splicing events.

Using the CIGAR string, the code constructs a splicing pattern representation by assigning "M" to represent included exons and "N" to represent skipped exons. It keeps track of the count for each unique splicing pattern observed.

The code provides insights into the specific splicing patterns present in the RNA-seq data, which can be helpful in studying alternative splicing events. 
"""

"""
"M" represents an included exon, indicating that the exon is present in the transcript produced from 
the alternative splicing event.

"N" represents a skipped exon, indicating that the exon is not included in the transcript produced from 
the alternative splicing event.
"""

bamfile = pysam.AlignmentFile("/content/test.bam", "rb")


splicing_patterns = {}

# Set the maximum number of reads to process couldn't exceed 100
max_reads = 90

for i, read in enumerate(bamfile):
    if read.is_unmapped:
        continue
    
    # Extract the cigar string from the read
    cigar = read.cigarstring
    

    if not cigar or "N" not in cigar:
        continue
    

    splicing_pattern = ""
    for op, length in read.cigartuples:
        if op == 0 or op == 2:
            splicing_pattern += length * "M"
        elif op == 3:
            splicing_pattern += length * "N"
    

    if splicing_pattern in splicing_patterns:
        splicing_patterns[splicing_pattern] += 1
    else:
        splicing_patterns[splicing_pattern] = 1
    

    if i >= max_reads:
        break


bamfile.close()

total_reads = sum(splicing_patterns.values())

# Calculate the frequencies of each splicing pattern
splicing_pattern_frequencies = {}
for splicing_pattern in splicing_patterns:
    if splicing_pattern in splicing_pattern_frequencies:
        splicing_pattern_frequencies[splicing_pattern] += 1
    else:
        splicing_pattern_frequencies[splicing_pattern] = 1


for splicing_pattern, frequency in splicing_pattern_frequencies.items():
    print("Splicing Pattern:", splicing_pattern)
    print("Frequency:", frequency)
    print()

"""# Perform Statistical Analysis for the cancer data"""

import numpy as np
# Calculate the frequencies and standard deviation
frequencies = np.array(list(splicing_pattern_frequencies.values())) / total_reads
std_deviation = np.std(frequencies)

# Print the results
print("Total Reads:", total_reads)
print("Number of Splicing Patterns:", len(splicing_pattern_frequencies))
print("Standard Deviation of Frequencies:", std_deviation)

"""# Convert the data that could be used to Matplotlib"""



"""# Visualize the data using Matplotlib"""



"""# Output the data in tsv of csv file"""

